File: taurus-protect-sdk-go/test/integration/whitelisting_test.go
package integration

import (
	"context"
	"encoding/json"
	"testing"

	"github.com/taurushq-io/taurus-protect-sdk/taurus-protect-sdk-go/pkg/protect/model"
)

func TestPaginateAllWhitelistedAddresses(t *testing.T) {
	skipIfNotIntegration(t)
	client := getTestClient(t)
	defer client.Close()

	ctx := context.Background()
	pageSize := int64(50)
	allAddresses := make([]*model.WhitelistedAddress, 0)
	offset := int64(0)

	// Fetch all whitelisted addresses using pagination
	for {
		addresses, pagination, err := client.WhitelistedAddresses().ListWhitelistedAddresses(ctx, &model.ListWhitelistedAddressesOptions{
			Limit:  pageSize,
			Offset: offset,
		})
		if err != nil {
			t.Fatalf("ListWhitelistedAddresses(offset=%d) error = %v", offset, err)
		}

		allAddresses = append(allAddresses, addresses...)
		t.Logf("Fetched %d whitelisted addresses (offset=%d)", len(addresses), offset)

		for _, a := range addresses {
			t.Logf("  WhitelistedAddress: ID=%s, Address=%s, Label=%s, Blockchain=%s", a.ID, a.Address, a.Label, a.Blockchain)
		}

		if pagination == nil || !pagination.HasMore {
			break
		}
		offset += pageSize

		// Safety limit for tests
		if offset > 2000 {
			t.Log("Stopping pagination test at 2000 items")
			break
		}
	}

	t.Logf("Total whitelisted addresses fetched via pagination: %d", len(allAddresses))
}

func TestListWhitelistedAddresses(t *testing.T) {
	skipIfNotIntegration(t)
	client := getTestClient(t)
	defer client.Close()

	ctx := context.Background()

	addresses, _, err := client.WhitelistedAddresses().ListWhitelistedAddresses(ctx, &model.ListWhitelistedAddressesOptions{
		Limit: 10,
	})
	if err != nil {
		t.Fatalf("ListWhitelistedAddresses error = %v", err)
	}

	t.Logf("Found %d whitelisted addresses", len(addresses))
	for _, a := range addresses {
		t.Logf("  %s/%s: %s", a.Blockchain, a.Network, a.Address)
	}
}

func TestPaginateAllWhitelistedAssets(t *testing.T) {
	skipIfNotIntegration(t)
	client := getTestClient(t)
	defer client.Close()

	ctx := context.Background()
	pageSize := int64(10)
	allAssets := make([]*model.WhitelistedAsset, 0)
	offset := int64(0)

	// Fetch all whitelisted assets using pagination
	for {
		assets, pagination, err := client.WhitelistedAssets().ListWhitelistedAssets(ctx, &model.ListWhitelistedAssetsOptions{
			Limit:  pageSize,
			Offset: offset,
		})
		if err != nil {
			t.Fatalf("ListWhitelistedAssets(offset=%d) error = %v", offset, err)
		}

		allAssets = append(allAssets, assets...)
		t.Logf("Fetched %d whitelisted assets (offset=%d)", len(assets), offset)

		for _, a := range allAssets[len(allAssets)-len(assets):] {
			t.Logf("  WhitelistedAsset: ID=%s, Blockchain=%s, Network=%s, Status=%s", a.ID, a.Blockchain, a.Network, a.Status)
		}

		if pagination == nil || !pagination.HasMore {
			break
		}
		offset += pageSize

		// Safety limit for tests
		if offset > 5000 {
			t.Log("Stopping pagination test at 5000 items")
			break
		}
	}

	t.Logf("Total whitelisted assets fetched via pagination: %d", len(allAssets))
}

func TestListWhitelistedAssets(t *testing.T) {
	skipIfNotIntegration(t)
	client := getTestClient(t)
	defer client.Close()

	ctx := context.Background()

	assets, _, err := client.WhitelistedAssets().ListWhitelistedAssets(ctx, &model.ListWhitelistedAssetsOptions{
		Limit: 10,
	})
	if err != nil {
		t.Fatalf("ListWhitelistedAssets error = %v", err)
	}

	t.Logf("Found %d whitelisted assets", len(assets))
	for _, a := range assets {
		t.Logf("  ID=%s, Blockchain=%s, Network=%s, Status=%s", a.ID, a.Blockchain, a.Network, a.Status)
	}
}

func TestGetWhitelistedAddress(t *testing.T) {
	skipIfNotIntegration(t)
	client := getTestClient(t)
	defer client.Close()

	ctx := context.Background()

	// First list addresses to find a valid ID
	addresses, _, err := client.WhitelistedAddresses().ListWhitelistedAddresses(ctx, &model.ListWhitelistedAddressesOptions{
		Limit: 1,
	})
	if err != nil {
		t.Fatalf("ListWhitelistedAddresses error = %v", err)
	}

	if len(addresses) == 0 {
		t.Skip("No whitelisted addresses available for testing")
	}

	// Get by ID
	addressID := addresses[0].ID
	address, err := client.WhitelistedAddresses().GetWhitelistedAddress(ctx, addressID)
	if err != nil {
		t.Fatalf("GetWhitelistedAddress(id=%s) error = %v", addressID, err)
	}

	t.Logf("GetWhitelistedAddress: ID=%s, Blockchain=%s, Network=%s, Address=%s, Label=%s, Status=%s",
		address.ID, address.Blockchain, address.Network, address.Address, address.Label, address.Status)
}

func TestListWhitelistedAddressesByBlockchain(t *testing.T) {
	skipIfNotIntegration(t)
	client := getTestClient(t)
	defer client.Close()

	ctx := context.Background()

	// List addresses filtered by blockchain ETH
	addresses, pagination, err := client.WhitelistedAddresses().ListWhitelistedAddresses(ctx, &model.ListWhitelistedAddressesOptions{
		Blockchain: "ETH",
		Limit:      10,
	})
	if err != nil {
		t.Fatalf("ListWhitelistedAddresses(blockchain=ETH) error = %v", err)
	}

	t.Logf("Listed %d whitelisted addresses for blockchain ETH", len(addresses))
	if pagination != nil {
		t.Logf("Pagination: TotalItems=%d, HasMore=%v", pagination.TotalItems, pagination.HasMore)
	}

	for _, a := range addresses {
		t.Logf("  WhitelistedAddress: ID=%s, Blockchain=%s, Network=%s, Address=%s",
			a.ID, a.Blockchain, a.Network, a.Address)
		// Verify filter is working
		if a.Blockchain != "ETH" {
			t.Errorf("Expected Blockchain=ETH, got %s", a.Blockchain)
		}
	}
}

func TestListWhitelistedAddressesByBlockchainAndNetwork(t *testing.T) {
	skipIfNotIntegration(t)
	client := getTestClient(t)
	defer client.Close()

	ctx := context.Background()

	// List addresses filtered by blockchain ETH and network mainnet
	addresses, pagination, err := client.WhitelistedAddresses().ListWhitelistedAddresses(ctx, &model.ListWhitelistedAddressesOptions{
		Blockchain: "ETH",
		Network:    "mainnet",
		Limit:      10,
	})
	if err != nil {
		t.Fatalf("ListWhitelistedAddresses(blockchain=ETH, network=mainnet) error = %v", err)
	}

	t.Logf("Listed %d whitelisted addresses for blockchain ETH, network mainnet", len(addresses))
	if pagination != nil {
		t.Logf("Pagination: TotalItems=%d, HasMore=%v", pagination.TotalItems, pagination.HasMore)
	}

	for _, a := range addresses {
		t.Logf("  WhitelistedAddress: ID=%s, Blockchain=%s, Network=%s, Address=%s",
			a.ID, a.Blockchain, a.Network, a.Address)
		// Verify filters are working
		if a.Blockchain != "ETH" {
			t.Errorf("Expected Blockchain=ETH, got %s", a.Blockchain)
		}
		if a.Network != "mainnet" {
			t.Errorf("Expected Network=mainnet, got %s", a.Network)
		}
	}
}

func TestGetWhitelistedAsset(t *testing.T) {
	skipIfNotIntegration(t)
	client := getTestClient(t)
	defer client.Close()

	ctx := context.Background()

	// First list assets to find a valid ID
	assets, _, err := client.WhitelistedAssets().ListWhitelistedAssets(ctx, &model.ListWhitelistedAssetsOptions{
		Limit: 1,
	})
	if err != nil {
		t.Fatalf("ListWhitelistedAssets error = %v", err)
	}

	if len(assets) == 0 {
		t.Skip("No whitelisted assets available for testing")
	}

	// Get by ID
	assetID := assets[0].ID
	asset, err := client.WhitelistedAssets().GetWhitelistedAsset(ctx, assetID)
	if err != nil {
		t.Fatalf("GetWhitelistedAsset(id=%s) error = %v", assetID, err)
	}

	t.Logf("GetWhitelistedAsset: ID=%s, Blockchain=%s, Network=%s, Status=%s",
		asset.ID, asset.Blockchain, asset.Network, asset.Status)

	// SECURITY: Log metadata payload fields by parsing from PayloadAsString (verified source)
	if asset.Metadata != nil && asset.Metadata.PayloadAsString != "" {
		var payload map[string]interface{}
		if err := json.Unmarshal([]byte(asset.Metadata.PayloadAsString), &payload); err == nil {
			if symbol, ok := payload["symbol"].(string); ok {
				t.Logf("  Symbol=%s", symbol)
			}
			if contractAddress, ok := payload["contractAddress"].(string); ok {
				t.Logf("  ContractAddress=%s", contractAddress)
			}
			if name, ok := payload["name"].(string); ok {
				t.Logf("  Name=%s", name)
			}
			if decimals, ok := payload["decimals"]; ok {
				t.Logf("  Decimals=%v", decimals)
			}
		}
	}
}

func TestListWhitelistedAssetsByBlockchain(t *testing.T) {
	skipIfNotIntegration(t)
	client := getTestClient(t)
	defer client.Close()

	ctx := context.Background()

	// List assets filtered by blockchain ETH
	assets, pagination, err := client.WhitelistedAssets().ListWhitelistedAssets(ctx, &model.ListWhitelistedAssetsOptions{
		Blockchain: "ETH",
		Limit:      10,
	})
	if err != nil {
		t.Fatalf("ListWhitelistedAssets(blockchain=ETH) error = %v", err)
	}

	t.Logf("Listed %d whitelisted assets for blockchain ETH", len(assets))
	if pagination != nil {
		t.Logf("Pagination: TotalItems=%d, HasMore=%v", pagination.TotalItems, pagination.HasMore)
	}

	for _, a := range assets {
		t.Logf("  WhitelistedAsset: ID=%s, Blockchain=%s, Network=%s, Status=%s",
			a.ID, a.Blockchain, a.Network, a.Status)
		// Verify filter is working
		if a.Blockchain != "ETH" {
			t.Errorf("Expected Blockchain=ETH, got %s", a.Blockchain)
		}
	}
}

func TestListWhitelistedAssetsByBlockchainAndNetwork(t *testing.T) {
	skipIfNotIntegration(t)
	client := getTestClient(t)
	defer client.Close()

	ctx := context.Background()

	// List assets filtered by blockchain ETH and network mainnet
	assets, pagination, err := client.WhitelistedAssets().ListWhitelistedAssets(ctx, &model.ListWhitelistedAssetsOptions{
		Blockchain: "ETH",
		Network:    "mainnet",
		Limit:      10,
	})
	if err != nil {
		t.Fatalf("ListWhitelistedAssets(blockchain=ETH, network=mainnet) error = %v", err)
	}

	t.Logf("Listed %d whitelisted assets for blockchain ETH, network mainnet", len(assets))
	if pagination != nil {
		t.Logf("Pagination: TotalItems=%d, HasMore=%v", pagination.TotalItems, pagination.HasMore)
	}

	for _, a := range assets {
		t.Logf("  WhitelistedAsset: ID=%s, Blockchain=%s, Network=%s, Status=%s",
			a.ID, a.Blockchain, a.Network, a.Status)
		// Verify filters are working
		if a.Blockchain != "ETH" {
			t.Errorf("Expected Blockchain=ETH, got %s", a.Blockchain)
		}
		if a.Network != "mainnet" {
			t.Errorf("Expected Network=mainnet, got %s", a.Network)
		}
	}
}

File: taurus-protect-sdk-typescript/tests/integration/whitelisting.test.ts
/**
 * Integration tests for Address and Asset Whitelisting APIs.
 *
 * These tests require a live API connection. Configure via environment variables
 * or hard-coded defaults in helpers.ts.
 *
 * Tests use high-level services (client.whitelistedAddresses, client.whitelistedAssets)
 * rather than raw OpenAPI APIs to demonstrate proper SDK usage patterns.
 */

import { skipIfNotIntegration, getTestClient } from "./helpers";

describe("Integration: Whitelisting", () => {
  beforeEach(() => {
    try {
      skipIfNotIntegration();
    } catch (e) {
      if (e instanceof Error && e.message === "SKIP_INTEGRATION") {
        return;
      }
      throw e;
    }
  });

  describe("Whitelisted Addresses", () => {
    it("listWhitelistedAddresses", async () => {
      try {
        skipIfNotIntegration();
      } catch (e) {
        if (e instanceof Error && e.message === "SKIP_INTEGRATION") {
          return;
        }
        throw e;
      }

      const client = getTestClient();
      try {
        const result = await client.whitelistedAddresses.list({ limit: 10 });

        const addresses = result.items;
        console.log(`Found ${addresses.length} whitelisted addresses`);

        if (result.pagination) {
          console.log(`Total items: ${result.pagination.totalItems}`);
        }

        for (const addr of addresses) {
          console.log(
            `  Address: ID=${addr.id}, Blockchain=${addr.blockchain}, Network=${addr.network}`
          );
        }

        expect(Array.isArray(addresses)).toBe(true);
      } finally {
        client.close();
      }
    });

    it(
      "paginateAllWhitelistedAddresses",
      async () => {
        try {
          skipIfNotIntegration();
        } catch (e) {
          if (e instanceof Error && e.message === "SKIP_INTEGRATION") {
            return;
          }
          throw e;
        }

        const client = getTestClient();
        try {
          const pageSize = 50;
          const allAddresses: Array<{ id?: string }> = [];
          let offset = 0;
          // Safety limit for addresses (aligned across all SDKs)
          const safetyLimit = 2000;

          // Fetch whitelisted addresses using pagination
          while (true) {
            const result = await client.whitelistedAddresses.list({
              limit: pageSize,
              offset: offset,
            });

            const addresses = result.items;
            allAddresses.push(...addresses);
            console.log(
              `Fetched ${addresses.length} whitelisted addresses (offset=${offset})`
            );

            // Check if there are more pages
            const totalItems = result.pagination?.totalItems ?? 0;
            if (offset + pageSize >= totalItems || addresses.length === 0) {
              break;
            }

            offset += pageSize;

            // Safety limit to avoid excessive API calls
            if (offset >= safetyLimit) {
              console.log(
                `Stopping pagination test at ${safetyLimit} items (safety limit)`
              );
              break;
            }
          }

          console.log(
            `Total whitelisted addresses fetched via pagination: ${allAddresses.length}`
          );
          expect(allAddresses.length).toBeGreaterThanOrEqual(0);
        } finally {
          client.close();
        }
      },
      300000
    );

    it("getWhitelistedAddress", async () => {
      try {
        skipIfNotIntegration();
      } catch (e) {
        if (e instanceof Error && e.message === "SKIP_INTEGRATION") {
          return;
        }
        throw e;
      }

      const client = getTestClient();
      try {
        // First, get a list to find a valid ID
        const listResult = await client.whitelistedAddresses.list({ limit: 1 });

        const addresses = listResult.items;
        if (addresses.length === 0) {
          console.log("No whitelisted addresses available for testing");
          return;
        }

        const addressId = addresses[0].id;
        if (!addressId) {
          console.log("Whitelisted address ID is undefined");
          return;
        }

        const address = await client.whitelistedAddresses.get(addressId);

        expect(address).toBeDefined();

        console.log("Whitelisted address details:");
        console.log(`  ID: ${address.id}`);
        console.log(`  Address: ${address.address}`);
        console.log(`  Blockchain: ${address.blockchain}`);
        console.log(`  Network: ${address.network}`);
      } finally {
        client.close();
      }
    });

    it("listWhitelistedAddressesByBlockchain", async () => {
      try {
        skipIfNotIntegration();
      } catch (e) {
        if (e instanceof Error && e.message === "SKIP_INTEGRATION") {
          return;
        }
        throw e;
      }

      const client = getTestClient();
      try {
        const result = await client.whitelistedAddresses.list({
          limit: 10,
          blockchain: "ETH",
        });

        const addresses = result.items;
        console.log(
          `Found ${addresses.length} whitelisted addresses filtered by blockchain=ETH`
        );

        if (result.pagination) {
          console.log(`Total items: ${result.pagination.totalItems}`);
        }

        for (const addr of addresses) {
          console.log(
            `  Address: ID=${addr.id}, Blockchain=${addr.blockchain}, Network=${addr.network}`
          );
        }

        expect(Array.isArray(addresses)).toBe(true);
        // All returned addresses should be ETH blockchain
        for (const addr of addresses) {
          expect(addr.blockchain).toBe("ETH");
        }
      } finally {
        client.close();
      }
    });

    it("listWhitelistedAddressesByBlockchainAndNetwork", async () => {
      try {
        skipIfNotIntegration();
      } catch (e) {
        if (e instanceof Error && e.message === "SKIP_INTEGRATION") {
          return;
        }
        throw e;
      }

      const client = getTestClient();
      try {
        const result = await client.whitelistedAddresses.list({
          limit: 10,
          blockchain: "ETH",
          network: "mainnet",
        });

        const addresses = result.items;
        console.log(
          `Found ${addresses.length} whitelisted addresses filtered by blockchain=ETH, network=mainnet`
        );

        if (result.pagination) {
          console.log(`Total items: ${result.pagination.totalItems}`);
        }

        for (const addr of addresses) {
          console.log(
            `  Address: ID=${addr.id}, Blockchain=${addr.blockchain}, Network=${addr.network}`
          );
        }

        expect(Array.isArray(addresses)).toBe(true);
        // Verify filtering worked (API returns filtered results)
        // Note: blockchain and network in the model come from the verified payload,
        // which may not include all fields. The API filtering works correctly,
        // but the model may have empty values if not in the signed payload.
        // This is intentional for security - we only expose verified data.
        expect(addresses.length).toBeGreaterThan(0);
        // Verify blockchain is set (comes from payload 'currency' field for addresses)
        for (const addr of addresses) {
          // blockchain may be empty if payload doesn't have 'currency' field
          // network may be empty if payload doesn't have 'network' field
          // We just verify we got results from the filtered API call
          expect(typeof addr.blockchain).toBe("string");
          expect(typeof addr.network).toBe("string");
        }
      } finally {
        client.close();
      }
    });
  });

  describe("Whitelisted Assets (Contracts)", () => {
    it("listWhitelistedAssets", async () => {
      try {
        skipIfNotIntegration();
      } catch (e) {
        if (e instanceof Error && e.message === "SKIP_INTEGRATION") {
          return;
        }
        throw e;
      }

      const client = getTestClient();
      try {
        const result = await client.whitelistedAssets.list({ limit: 10 });

        const assets = result.items;
        console.log(`Found ${assets.length} whitelisted assets (contracts)`);

        if (result.pagination) {
          console.log(`Total items: ${result.pagination.totalItems}`);
        }

        for (const asset of assets) {
          console.log(
            `  Asset: ID=${asset.id}, Blockchain=${asset.blockchain}, Symbol=${asset.symbol}`
          );
        }

        expect(Array.isArray(assets)).toBe(true);
      } finally {
        client.close();
      }
    });

    it(
      "paginateAllWhitelistedAssets",
      async () => {
        try {
          skipIfNotIntegration();
        } catch (e) {
          if (e instanceof Error && e.message === "SKIP_INTEGRATION") {
            return;
          }
          throw e;
        }

        const client = getTestClient();
        try {
          const pageSize = 100;
          const allAssets: Array<{ id?: number }> = [];
          let offset = 0;
          const safetyLimit = 5000;

          // Fetch whitelisted assets using pagination
          while (true) {
            const result = await client.whitelistedAssets.list({
              limit: pageSize,
              offset: offset,
            });

            const assets = result.items;
            allAssets.push(...assets);
            console.log(
              `Fetched ${assets.length} whitelisted assets (offset=${offset})`
            );

            // Check if there are more pages
            const totalItems = result.pagination?.totalItems ?? 0;
            if (offset + pageSize >= totalItems || assets.length === 0) {
              break;
            }

            offset += pageSize;

            // Safety limit to avoid excessive API calls
            if (offset >= safetyLimit) {
              console.log(
                `Stopping pagination test at ${safetyLimit} items (safety limit)`
              );
              break;
            }
          }

          console.log(
            `Total whitelisted assets fetched via pagination: ${allAssets.length}`
          );
          expect(allAssets.length).toBeGreaterThanOrEqual(0);
        } finally {
          client.close();
        }
      },
      60000
    );

    it("getWhitelistedAsset", async () => {
      try {
        skipIfNotIntegration();
      } catch (e) {
        if (e instanceof Error && e.message === "SKIP_INTEGRATION") {
          return;
        }
        throw e;
      }

      const client = getTestClient();
      try {
        // First, get a list to find a valid ID
        const listResult = await client.whitelistedAssets.list({ limit: 1 });

        const assets = listResult.items;
        if (assets.length === 0) {
          console.log("No whitelisted assets available for testing");
          return;
        }

        const assetId = assets[0].id;
        if (!assetId) {
          console.log("Whitelisted asset ID is undefined");
          return;
        }

        const asset = await client.whitelistedAssets.get(assetId);

        expect(asset).toBeDefined();

        console.log("Whitelisted asset details:");
        console.log(`  ID: ${asset.id}`);
        console.log(`  Contract: ${asset.contractAddress}`);
        console.log(`  Blockchain: ${asset.blockchain}`);
        console.log(`  Network: ${asset.network}`);
      } finally {
        client.close();
      }
    });

    it("listWhitelistedAssetsByBlockchain", async () => {
      try {
        skipIfNotIntegration();
      } catch (e) {
        if (e instanceof Error && e.message === "SKIP_INTEGRATION") {
          return;
        }
        throw e;
      }

      const client = getTestClient();
      try {
        const result = await client.whitelistedAssets.list({
          limit: 10,
          blockchain: "ETH",
        });

        const assets = result.items;
        console.log(
          `Found ${assets.length} whitelisted assets filtered by blockchain=ETH`
        );

        if (result.pagination) {
          console.log(`Total items: ${result.pagination.totalItems}`);
        }

        for (const asset of assets) {
          console.log(
            `  Asset: ID=${asset.id}, Blockchain=${asset.blockchain}, Symbol=${asset.symbol}`
          );
        }

        expect(Array.isArray(assets)).toBe(true);
        // All returned assets should be ETH blockchain
        for (const asset of assets) {
          expect(asset.blockchain).toBe("ETH");
        }
      } finally {
        client.close();
      }
    });

    it("listWhitelistedAssetsByBlockchainAndNetwork", async () => {
      try {
        skipIfNotIntegration();
      } catch (e) {
        if (e instanceof Error && e.message === "SKIP_INTEGRATION") {
          return;
        }
        throw e;
      }

      const client = getTestClient();
      try {
        const result = await client.whitelistedAssets.list({
          limit: 10,
          blockchain: "ETH",
          network: "mainnet",
        });

        const assets = result.items;
        console.log(
          `Found ${assets.length} whitelisted assets filtered by blockchain=ETH, network=mainnet`
        );

        if (result.pagination) {
          console.log(`Total items: ${result.pagination.totalItems}`);
        }

        for (const asset of assets) {
          console.log(
            `  Asset: ID=${asset.id}, Blockchain=${asset.blockchain}, Network=${asset.network}, Symbol=${asset.symbol}`
          );
        }

        expect(Array.isArray(assets)).toBe(true);
        // Verify filtering worked (API returns filtered results)
        // Note: blockchain and network in the model come from the verified payload,
        // which may not include all fields. The API filtering works correctly,
        // but the model may have empty values if not in the signed payload.
        // This is intentional for security - we only expose verified data.
        expect(assets.length).toBeGreaterThan(0);
        // Verify blockchain is set (comes from payload 'blockchain' field for assets)
        for (const asset of assets) {
          // blockchain may be empty if payload doesn't have 'blockchain' field
          // network may be empty if payload doesn't have 'network' field
          // We just verify we got results from the filtered API call
          expect(typeof asset.blockchain).toBe("string");
          expect(typeof asset.network).toBe("string");
        }
      } finally {
        client.close();
      }
    });
  });
});
